File: cpp/19-remove-nth-node-from-end-of-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode *p = head;
        vector<ListNode*> nodes;
        int size = 0;
        
        while(p){
            nodes.push_back(p);
            p = p->next;
        }
        
        size = nodes.size();
        if(size > n){
            p = nodes[size - n - 1];
            p->next = p->next->next;
            p = nodes[size - n];
            delete(p);
        }
        else{
            p = head;
            head = head->next;
            delete(p);
        }

        return head;
    }
};

File: py/19-remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        fast = slow = head
        
        for _ in range(n):
            fast = fast.next
            
        if fast:
            while fast.next:
                fast = fast.next
                slow = slow.next
        else:
            return head.next
        
        slow.next = slow.next.next
            
        return head